lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/ManagedStrategyTransformer.java

```diff
@@ -48,80 +46,659 @@
+    private static final Method IS_LOCAL_OBJECT_METHOD = new Method("isLocalObject", Type.BOOLEAN_TYPE, new Type[]{OBJECT_TYPE});
+    private static final Method ADD_DEPENDENCY = new Method("addDependency", Type.VOID_TYPE, new Type[]{OBJECT_TYPE, OBJECT_TYPE});
+
+
+    private static final Method GET_UNSAFE_METHOD = new Method("getUnsafe", UNSAFE_TYPE, new Type[]{});
+    private static final Method CLASS_FOR_NAME = new Method("forName", CLASS_TYPE, new Type[]{STRING_TYPE});
 
     private String className;
+    private int classVersion;
     private String fileName;
-    private final List<StackTraceElement> codeLocations = new ArrayList<>();
+    private final List<StackTraceElement> codeLocations;
 
-    public ManagedStrategyTransformer(ClassVisitor cv, ClassInfo ci) {
-        super(ASM_API, cv);
+    ManagedStrategyTransformer(ClassVisitor cv, ClassInfo ci, List<StackTraceElement> codeLocations) {
+        super(ASM_API, new ClassRemapper(cv, new JavaUtilRemapper()));
         this.className = ci.getClassName();
+        this.codeLocations = codeLocations;
+        this.classVersion = ci.getVersion();
         this.fileName = ci.getSourceFile();
     }
 
-    public List<StackTraceElement> getCodeLocations() {
+    List<StackTraceElement> getCodeLocations() {
         return codeLocations;
     }
 
     @Override
     public MethodVisitor visitMethod(int access, String mname, String desc, String signature, String[] exceptions) {
+        //System.out.println(className);
+        // replace native method VMSupportsCS8 in AtomicLong with stub
+        if ((access & ACC_NATIVE) != 0 && mname.equals("VMSupportsCS8")) {
+            MethodVisitor mv = super.visitMethod(access & ~ACC_NATIVE, mname, desc, signature, exceptions);
+            return new NativeMethodStubTransformer(new GeneratorAdapter(mv, access & ~ACC_NATIVE, mname, desc));
+        }
+
+        boolean isSynchronized = (access & ACC_SYNCHRONIZED) != 0;
+        if (isSynchronized) {
+            access ^= ACC_SYNCHRONIZED; // disable synchronized
+        }
+
         MethodVisitor mv = super.visitMethod(access, mname, desc, signature, exceptions);
         mv = new JSRInlinerAdapter(mv, access, mname, desc, signature, exceptions);
+        mv = new IndirectSharedVariableAccessMethodTransformer(mname, new GeneratorAdapter(mv, access, mname, desc));
+        mv = new SynchronizedLockTransformer(mname, new GeneratorAdapter(mv, access, mname, desc));
+
+        if (isSynchronized) {
+            // synchronized method is replaced with synchronized lock
+            mv = new SynchronizedLockAddingTransformer(mname, new GeneratorAdapter(mv, access, mname, desc), className, access, classVersion);
+        }
+
+        if ("<clinit>".equals(mname)) {
+            mv = new ClassInitializationTransformer(mname, new GeneratorAdapter(mv, access, mname, desc));
+        }
+
+        mv = new WaitNotifyTransformer(mname, new GeneratorAdapter(mv, access, mname, desc));
```

Comment:
Do you need to apply all these transformers to <clinit>?

Reply:
To what?